File: C1-Algorithmic_Toolbox/W2-Algorithmic_Warmup/5_fibonacci_number_again/fibonacci_huge.py
# python3


def fibonacci_huge_naive(n, m):
    # Naive solution:
    if n <= 1:
        return n
    previous = 0
    current = 1
    for _ in range(n - 1):
        previous, current = current, previous + current
    return current % m


def get_pisano_period(m):
    a = 0
    b = 1
    c = a + b
    for i in range(m * m):
        c = (a + b) % m
        a = b
        b = c
        if a == 0 and b == 1:
            return i + 1


def fibonacci_huge_fast(n, m):
    # Faster solution:
    pisano_period = get_pisano_period(m)
    remainder = n % pisano_period
    # print("pisano_period: ", pisano_period)
    # print("remainder: ", remainder)

    if remainder <= 1:
        return remainder

    prev = 0
    curr = 1
    temp = 0

    for _ in range(2, remainder + 1):
        temp = curr
        curr = (prev + curr) % m
        prev = temp
    return curr % m


if __name__ == "__main__":
    n, m = map(int, input().split())
    print(fibonacci_huge_fast(n, m))

File: C1-Algorithmic_Toolbox/W2-Algorithmic_Warmup/5_fibonacci_number_again/fibonacci_huge.cpp
#include <iostream>

using namespace std;

long long get_fibonacci_huge_naive(long long n, long long m) {
    if (n <= 1) return n;

    long long previous = 0;
    long long current = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
    }

    return current % m;
}

long long find_Pisano_period(long long m) {
    long long prev = 0, curr = 1, temp = 0;
    long long result = 0;
    for (int i = 0; i < m * m; ++i) {
        temp = curr;
        curr = (prev + curr) % m;
        prev = temp;
        if (prev == 0 && curr == 1) {
            result = i + 1;
        }
    }

    return result;
}

long long get_fibonacci_huge_fast(long long n, long long m) {
    long long pisano_period = find_Pisano_period(m);
    long long remainder = n % pisano_period;
    // cout << "pisano_period: " << pisano_period << endl;
    // cout << "remainder: " << remainder << endl;

    if (remainder <= 1) return remainder;

    long long prev = 0, curr = 1, temp = 0;
    for (long long i = 2; i < remainder + 1; ++i) {
        temp = curr;
        curr = (prev + curr) % m;
        prev = temp;
    }

    return curr % m;
}

// g++ -pipe -O2 -std=c++14 fibonacci_huge.cpp -lm && ./a.out
int main() {
    long long n, m;
    cin >> n >> m;
    cout << get_fibonacci_huge_fast(n, m) << '\n';
}
